tests/Configuration/RawConfiguration/RawConfigurationTest.php

```diff
@@ -0,0 +1,179 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\Configuration\RawConfiguration;
+
+use Exception;
+use Generator;
+use Infection\Configuration\RawConfiguration\InvalidFile;
+use Infection\Configuration\RawConfiguration\RawConfiguration;
+use PHPUnit\Framework\TestCase;
+use ReflectionClass;
+use Seld\JsonLint\ParsingException;
+
+final class RawConfigurationTest extends TestCase
+{
+    private const FIXTURES_DIR = __DIR__ . '/../../Fixtures/Configuration';
+
+    public function test_it_can_be_instantiated(): void
+    {
+        $path = '/nowhere';
+
+        $config = new RawConfiguration($path);
+
+        $this->assertSame($path, $config->getPath());
+    }
+
+    public function test_its_contents_is_retrieved_lazily(): void
+    {
+        $invalidPath = '/nowhere';
+
+        $config = new RawConfiguration($invalidPath);
+
+        try {
+            $config->getDecodedContents();
+
+            $this->fail('Expected the content to be invalid.');
+        } catch (Exception $exception) {
+            $this->addToAssertionCount(1);
+        }
+
+        $validPath = self::FIXTURES_DIR . '/file.json';
+        $expectedArrayContents = ['foo' => 'bar'];
+
+        $config = new RawConfiguration($validPath);
+        $actualContents = $config->getDecodedContents();
+
+        $this->assertSame($expectedArrayContents, (array) $actualContents);
+    }
+
+    public function test_its_contents_is_retrieved_only_once(): void
+    {
+        $config = new RawConfiguration(self::FIXTURES_DIR . '/file.json');
+        $expectedValue = (object) ['a' => 'b'];
+
+        // Fetch the contents once
+        $config->getDecodedContents();
+
+        $decodedContentsReflection = (new ReflectionClass(RawConfiguration::class))->getProperty('decodedContents');
+        $decodedContentsReflection->setAccessible(true);
+        $decodedContentsReflection->setValue($config, $expectedValue);
+
+        $this->assertSame($expectedValue, $config->getDecodedContents());
+    }
+
+    /**
+     * @dataProvider invalidConfigContentsProvider
+     */
+    public function test_it_cannot_retrieve_or_decode_invalid_contents(
+        string $path,
+        Exception $expectedException
+    ): void {
+        $config = new RawConfiguration($path);
+
+        try {
+            $config->getDecodedContents();
+
+            $this->fail('Expected the config contents to be invalid.');
```

Comment:
Why aren't we using `exceptException` here?

Reply:
I really tend to avoid `exceptException` overall since it really makes it confusing what block is supposed to throw. It also makes harder to have assertions against the thrown exception like we have here